# Exploring eBay Car Sales Data


This dataset contains info about 50,000 cars in sale, and is available by this [link.](https://data.world/data-society/used-cars-data)
<br>It contains some columns, such as dateCrawled, ...
<br> The aim of this project is to clean and analyze this cars' list.

In [3]:
import pandas as pd
import numpy as np

In [4]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

Jupyter Notebook shows the first and last 5 rows itself, by typing the mere "autos" 

In [5]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

From the general observation, what we see is that some columns, such as _vehicleType, gearbox, model, fuelType, notRepairedDamage_ have some rows with no values. 
<br>Also, 5 columns have integer values (_yearOfRegistration, powerPS, monthOfRegistration, nrOfPictures, postalCode_).

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [7]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

One thing to mention: column names in this dataset contain camelCase instead of snake_case, which is preferred in Python. 

In [8]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [9]:
new_col = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [10]:
col_dict = {}
for i in range(len(new_col)):
    col_dict[autos.columns[i]] = new_col[i]
col_dict

{'dateCrawled': 'date_crawled',
 'name': 'name',
 'seller': 'seller',
 'offerType': 'offer_type',
 'price': 'price',
 'abtest': 'abtest',
 'vehicleType': 'vehicle_type',
 'yearOfRegistration': 'registration_year',
 'gearbox': 'gearbox',
 'powerPS': 'power_ps',
 'model': 'model',
 'odometer': 'odometer',
 'monthOfRegistration': 'registration_month',
 'fuelType': 'fuel_type',
 'brand': 'brand',
 'notRepairedDamage': 'unrepaired_damage',
 'dateCreated': 'ad_created',
 'nrOfPictures': 'nr_of_pictures',
 'postalCode': 'postal_code',
 'lastSeen': 'last_seen'}

In [11]:
autos.rename(columns=col_dict, inplace=True)

In [12]:
autos.head()
# We made the columns of autos easier to read 

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [13]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the given description above:
<br>1. _seller, offer_type_ columns have values 1 vs. 49999. 
<br>2. Also, _abtest, gearbox, unrepaired_damage_ columns have 2 values only. Thus, I guess, it can be changed to binary 0 and 1.
<br>3. _registration_year_ column has max value of 9999 and min value of 1000, which is impossible. To add, _power_ps_ column has a max of 17700 and min of 0. _registration_month_ column has min of 0. 
<br>4.We can convert _price_ and _odometer_ columns to float.
##### We are completing the 4th point.

In [14]:
autos["price"] = autos["price"].str.replace('$','').str.replace(',','').astype(float)

In [15]:
autos["odometer"] = autos["odometer"].str.replace('km','').str.replace(',','').astype(float)

In [16]:
autos["odometer"]

0        150000.0
1        150000.0
2         70000.0
3         70000.0
4        150000.0
           ...   
49995    100000.0
49996    150000.0
49997      5000.0
49998     40000.0
49999    150000.0
Name: odometer, Length: 50000, dtype: float64

In [17]:
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

##### Now, let's search for outliers in values and clean the data.

In [18]:
autos['odometer_km'].unique().shape

(13,)

In [19]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos['odometer_km'].value_counts().sort_index()

5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64

##### From my point of perspective, there are no outliers in odometer_km.

In [21]:
autos['price'].unique().shape
pd.set_option('display.float_format', lambda x: '%.1f' % x)

In [22]:
autos['price'].describe()

count      50000.0
mean        9840.0
std       481104.4
min            0.0
25%         1100.0
50%         2950.0
75%         7200.0
max     99999999.0
Name: price, dtype: float64

##### Min price is 0 and Max price is 99999999, which are certainly impossible in eBay Sales. Thus, we need to remove those rows.

I guess, no one is going to sell his/her car for free, for a such cheap or expensive price. My boundary for car price is between 1,000 and 100,000 US dollars. This is a subjective opinion and boundaries can be changed easily.

In [23]:
LOWER_BOUND = 1000
UPPER_BOUND = 100001
autos['price'].value_counts().sort_index()

0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
              ... 
10000000.0       1
11111111.0       2
12345678.0       3
27322222.0       1
99999999.0       1
Name: price, Length: 2357, dtype: int64

In [24]:
autos = autos[autos['price'].between(LOWER_BOUND, UPPER_BOUND)]

In [25]:
autos['price'].value_counts().sort_index()

1000.0     639
1039.0       1
1040.0       1
1049.0       6
1050.0      95
          ... 
93911.0      1
94999.0      1
98500.0      1
99000.0      2
99900.0      2
Name: price, Length: 2052, dtype: int64

In [26]:
autos.describe()

price  registration_year  power_ps  odometer_km  registration_month  \
count 38587.0            38587.0   38587.0      38587.0             38587.0   
mean   7089.0             2005.7     128.7     122859.8                 6.0   
std    7893.8               86.7     215.8      40712.6                 3.6   
min    1000.0             1000.0       0.0       5000.0                 0.0   
25%    2200.0             2001.0      80.0     100000.0                 3.0   
50%    4300.0             2005.0     116.0     150000.0                 6.0   
75%    8950.0             2009.0     160.0     150000.0                 9.0   
max   99900.0             9999.0   17700.0     150000.0                12.0   

       nr_of_pictures  postal_code  
count         38587.0      38587.0  
mean              0.0      51682.9  
std               0.0      25689.2  
min               0.0       1067.0  
25%               0.0      31199.0  
50%               0.0      50829.0  
75%               0.0      72528.0  
max               0.0      99998.0

In [27]:
autos['odometer_km'].value_counts().sort_index()

5000.0        498
10000.0       225
20000.0       684
30000.0       742
40000.0       794
50000.0       984
60000.0      1097
70000.0      1154
80000.0      1333
90000.0      1569
100000.0     1858
125000.0     4338
150000.0    23311
Name: odometer_km, dtype: int64

### Now, let's go through date columns.

In [28]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [29]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 38587, dtype: object


In [30]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index() 

2016-03-05   0.0
2016-03-06   0.0
2016-03-07   0.0
2016-03-08   0.0
2016-03-09   0.0
2016-03-10   0.0
2016-03-11   0.0
2016-03-12   0.0
2016-03-13   0.0
2016-03-14   0.0
2016-03-15   0.0
2016-03-16   0.0
2016-03-17   0.0
2016-03-18   0.0
2016-03-19   0.0
2016-03-20   0.0
2016-03-21   0.0
2016-03-22   0.0
2016-03-23   0.0
2016-03-24   0.0
2016-03-25   0.0
2016-03-26   0.0
2016-03-27   0.0
2016-03-28   0.0
2016-03-29   0.0
2016-03-30   0.0
2016-03-31   0.0
2016-04-01   0.0
2016-04-02   0.0
2016-04-03   0.0
2016-04-04   0.0
2016-04-05   0.0
2016-04-06   0.0
2016-04-07   0.0
Name: date_crawled, dtype: float64

In [31]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index() 

2015-06-11   0.0
2015-08-10   0.0
2015-09-09   0.0
2015-11-10   0.0
2015-12-30   0.0
              ..
2016-04-03   0.0
2016-04-04   0.0
2016-04-05   0.0
2016-04-06   0.0
2016-04-07   0.0
Name: ad_created, Length: 74, dtype: float64

In [32]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index() 

2016-03-05   0.0
2016-03-06   0.0
2016-03-07   0.0
2016-03-08   0.0
2016-03-09   0.0
2016-03-10   0.0
2016-03-11   0.0
2016-03-12   0.0
2016-03-13   0.0
2016-03-14   0.0
2016-03-15   0.0
2016-03-16   0.0
2016-03-17   0.0
2016-03-18   0.0
2016-03-19   0.0
2016-03-20   0.0
2016-03-21   0.0
2016-03-22   0.0
2016-03-23   0.0
2016-03-24   0.0
2016-03-25   0.0
2016-03-26   0.0
2016-03-27   0.0
2016-03-28   0.0
2016-03-29   0.0
2016-03-30   0.0
2016-03-31   0.0
2016-04-01   0.0
2016-04-02   0.0
2016-04-03   0.0
2016-04-04   0.0
2016-04-05   0.1
2016-04-06   0.2
2016-04-07   0.1
Name: last_seen, dtype: float64

From the date columns, what I see is that the data is mostly gathered in 2016, it helps us for the registration_year column.

In [33]:
autos["registration_year"].describe()

count   38587.0
mean     2005.7
std        86.7
min      1000.0
25%      2001.0
50%      2005.0
75%      2009.0
max      9999.0
Name: registration_year, dtype: float64

In [34]:
LOWER_BOUND = 1980
UPPER_BOUND = 2016
autos = autos[autos['registration_year'].between(LOWER_BOUND, UPPER_BOUND)]
autos['registration_year'].describe()

count   36717.0
mean     2004.4
std         5.8
min      1980.0
25%      2001.0
50%      2005.0
75%      2008.0
max      2016.0
Name: registration_year, dtype: float64

Surely, data is gathered in 2016, thus the maximum year can be 2016. Also, I took the minimum of 1980's as an instance, because it can't be that we have a car registered in 1000s. 

## Next Section is Data Aggregation

We have 36717 cars of 40 brands left so far. I decided to choose top 20 brands of them.

In [35]:
autos['brand'].describe()

count          36717
unique            40
top       volkswagen
freq            7782
Name: brand, dtype: object

In [36]:
data_aggregation = {}

In [37]:
autos['brand'].value_counts()

volkswagen        7782
bmw               4639
mercedes_benz     4072
audi              3628
opel              3278
ford              2129
renault           1382
peugeot           1037
fiat               762
skoda              707
seat               643
smart              618
toyota             542
mazda              530
citroen            509
nissan             507
mini               403
hyundai            400
volvo              326
sonstige_autos     297
kia                286
honda              272
mitsubishi         256
porsche            240
chevrolet          229
alfa_romeo         224
suzuki             213
dacia              122
chrysler           115
jeep               101
land_rover          95
jaguar              67
subaru              64
daihatsu            63
saab                51
daewoo              34
rover               27
lancia              24
trabant             23
lada                20
Name: brand, dtype: int64

In [38]:
brands = autos['brand'].value_counts().index
new_brands = brands[:20]

In [39]:
for brand in new_brands:
    series = autos[autos['brand'] == brand]
    mean_series = series['price'].mean()
    data_aggregation[brand] = mean_series

In [40]:
k = 0
new_data = {}
for key in data_aggregation:
    if key in new_brands:
        new_data[key] = data_aggregation[key]
new_data

{'volkswagen': 6597.806604985864,
 'bmw': 8970.746281526192,
 'mercedes_benz': 9063.782907662082,
 'audi': 10273.969128996692,
 'opel': 4192.410921293472,
 'ford': 4908.70032879286,
 'renault': 3514.8241678726486,
 'peugeot': 3948.3760848601737,
 'fiat': 3923.968503937008,
 'skoda': 6841.9066478076375,
 'seat': 5638.640746500778,
 'smart': 3780.4692556634304,
 'toyota': 5567.477859778598,
 'mazda': 5309.526415094339,
 'citroen': 4493.8939096267195,
 'nissan': 6428.428007889546,
 'mini': 10724.992555831266,
 'hyundai': 6181.8725,
 'volvo': 6075.19018404908,
 'sonstige_autos': 12182.794612794612}

In [41]:
sorted_data = sorted(new_data.items(), key=lambda val: val[1])
sorted_data.reverse()
mean_prices = {}
for data in sorted_data:
    mean_prices[data[0]] = data[1]
mean_prices

{'sonstige_autos': 12182.794612794612,
 'mini': 10724.992555831266,
 'audi': 10273.969128996692,
 'mercedes_benz': 9063.782907662082,
 'bmw': 8970.746281526192,
 'skoda': 6841.9066478076375,
 'volkswagen': 6597.806604985864,
 'nissan': 6428.428007889546,
 'hyundai': 6181.8725,
 'volvo': 6075.19018404908,
 'seat': 5638.640746500778,
 'toyota': 5567.477859778598,
 'mazda': 5309.526415094339,
 'ford': 4908.70032879286,
 'citroen': 4493.8939096267195,
 'opel': 4192.410921293472,
 'peugeot': 3948.3760848601737,
 'fiat': 3923.968503937008,
 'smart': 3780.4692556634304,
 'renault': 3514.8241678726486}

## Conclusion
From the _sorted_data_ dictionary, it can be seen that the mean price of the top-20 put on is between 3500 and 12200 euros.
Sonstige_autos and Mini are expensive, because they are rare cars. Surprisingly, top-20 french cars are cheaper than others.
Audi, BMW, and Mercedes are expensive, while Ford and Opel are much more cheaper. Most abundant brand - Volskwagen is in between.

In [42]:
bmp_series = pd.Series(mean_prices)
print(bmp_series)

sonstige_autos   12182.8
mini             10725.0
audi             10274.0
mercedes_benz     9063.8
bmw               8970.7
skoda             6841.9
volkswagen        6597.8
nissan            6428.4
hyundai           6181.9
volvo             6075.2
seat              5638.6
toyota            5567.5
mazda             5309.5
ford              4908.7
citroen           4493.9
opel              4192.4
peugeot           3948.4
fiat              3924.0
smart             3780.5
renault           3514.8
dtype: float64


In [43]:
odom_means = {}
for brand in new_brands:
    odom = autos[autos['brand'] == brand]
    mean_odom = odom['odometer_km'].mean()
    odom_means[brand] = mean_odom
odom_means

{'volkswagen': 125970.83012079156,
 'bmw': 132234.31774089244,
 'mercedes_benz': 130424.85265225933,
 'audi': 127531.69790518192,
 'opel': 124385.29591214155,
 'ford': 120807.89102865195,
 'renault': 121711.28798842257,
 'peugeot': 122410.80038572806,
 'fiat': 108772.96587926509,
 'skoda': 110297.02970297029,
 'seat': 116104.19906687403,
 'smart': 97775.08090614887,
 'toyota': 113782.28782287824,
 'mazda': 119981.1320754717,
 'citroen': 114341.84675834971,
 'nissan': 110335.3057199211,
 'mini': 88933.00248138957,
 'hyundai': 101862.5,
 'volvo': 138282.20858895706,
 'sonstige_autos': 97154.88215488216}

In [45]:
bmo_series = pd.Series(odom_means)
bmo_series

volkswagen       125970.8
bmw              132234.3
mercedes_benz    130424.9
audi             127531.7
opel             124385.3
ford             120807.9
renault          121711.3
peugeot          122410.8
fiat             108773.0
skoda            110297.0
seat             116104.2
smart             97775.1
toyota           113782.3
mazda            119981.1
citroen          114341.8
nissan           110335.3
mini              88933.0
hyundai          101862.5
volvo            138282.2
sonstige_autos    97154.9
dtype: float64

In [51]:
bmp_df = pd.DataFrame(bmp_series, columns=['mean_price'])

In [52]:
bmp_df

mean_price
sonstige_autos     12182.8
mini               10725.0
audi               10274.0
mercedes_benz       9063.8
bmw                 8970.7
skoda               6841.9
volkswagen          6597.8
nissan              6428.4
hyundai             6181.9
volvo               6075.2
seat                5638.6
toyota              5567.5
mazda               5309.5
ford                4908.7
citroen             4493.9
opel                4192.4
peugeot             3948.4
fiat                3924.0
smart               3780.5
renault             3514.8

In [55]:
bmp_df['odometer_mean'] = bmo_series

In [56]:
bmp_df

mean_price  odometer_mean
sonstige_autos     12182.8        97154.9
mini               10725.0        88933.0
audi               10274.0       127531.7
mercedes_benz       9063.8       130424.9
bmw                 8970.7       132234.3
skoda               6841.9       110297.0
volkswagen          6597.8       125970.8
nissan              6428.4       110335.3
hyundai             6181.9       101862.5
volvo               6075.2       138282.2
seat                5638.6       116104.2
toyota              5567.5       113782.3
mazda               5309.5       119981.1
ford                4908.7       120807.9
citroen             4493.9       114341.8
opel                4192.4       124385.3
peugeot             3948.4       122410.8
fiat                3924.0       108773.0
smart               3780.5        97775.1
renault             3514.8       121711.3

From the table above, it can be seen that __sonstige_autos__ and __mini__ are the most expensive, and the one of the main reasons that their mean mileage is low (means they are newer than other cars). 

On the other hand, from the top-6 most posted cars, __audi__, __mercedes__, and __bmw__ are expensive, despite their mean mileage are much higher than previous two car brands. 

All of the top-6 brands have the mean mileage of __120000__ km at least. 

In [57]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type   price   abtest vehicle_type  registration_year  \
0      privat    Angebot  5000.0  control          bus               2004   
1      privat    Angebot  8500.0  control    limousine               1997   
2      privat    Angebot  8990.0     test    limousine               2009   
3      privat    Angebot  4350.0  control   kleinwagen               2007   
4      privat    Angebot  1350.0     test        kombi               2003   
...       ...        ...     ...      ...          ...                ...   
49995  privat    Angebot 24900.0  control    limousine               2011   
49996  privat    Angebot  1980.0  control       cabrio               1996   
49997  privat    Angebot 13200.0     test       cabrio               2014   
49998  privat    Angebot 22900.0  control        kombi               2013   
49999  privat    Angebot  1250.0  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere     150000.0                   3       lpg   
1      automatik       286     7er     150000.0                   6    benzin   
2        manuell       102    golf      70000.0                   7    benzin   
3      automatik        71  fortwo      70000.0                   6    benzin   
4        manuell         0   focus     150000.0                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5     100000.0                   1    diesel   
49996    manuell        75   astra     150000.0                   5    benzin   
49997  automatik        69     500       5000.0                  11    benzin   
49998    manuell       150      a3      40000.0                  11    diesel   
49999    manuell       101  vectra     150000.0                   1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  